Refactored MobileInfo, Mobile and Move

Rydra opened the pull request:
Refactored MobileInfo, Mobile and Move to make them testable in isolation, easier to understand what they do by commenting and restructuring the code making sure to not cause undesired side-effects.

The fact of having to break dependencies on a highly coupled system forced me to modify and create several interfaces, that at the same time forced me to change several classes directly and indirectly related to Move, Mobile and MobileInfo.

Additionally, I added several unit tests to try how testable they've become.

I've been comenting throughoutfully the code, I've been manipulating, yet to sum up:
- By no means this refactor is complete. It should be completed through the span of several PR. If you take a look at the MoveTests, you'll see that I had to mock a load of dependencies in order to make the tests pass successfully. This is a clear smell of Demeter violation. The construction of Move instances should be delegated to some kind of Move factory that can resolve all the move dependencies for him. This, utimately, will make Move easier to undertand.
- There is no clear barrier and meaning between Mobile and MobileInfo. There are lot of functions that Mobile could deal by himself and yet are delegated to MobileInfo...
- The crush, block, enterCell, finishmove notifications... are implemented in a way that couples actors to each other. We should think in a way to make this work on a subscribe/publish like the Observer Pattern: http://dotnetcodr.com/2013/08/01/design-patterns-and-practices-in-net-the-observer-pattern/ . A message bus implementation of this would be really smart for the future, and would skyroof the performance.

Rydra commented:
It seems there are some problems when downloading the dependencies. Does anyone have any clue about that?

LipkeGu commented:
Will this PR also fix the Problem that the PF tries to calculate paths which are never reachable? =)

Rydra commented:
@LipkeGu Nope, but it will make it way easier to track down the problem by implementing a unit test that expresses this behaviour and it's what it is expected. The test initially should fail, and then work it out until we make it pass.

One idea should be, first, manage to get rid of the unholy FindBidiPath, then add a maxDepth to limit the search area of the pathfinder.

This is comented as an improvement through the code.

LipkeGu commented:
Maybe we should also filter by directions when we have a path to the target (in front of the unit we doesnt need to calculate the ways behind the Unit except it is not able to find a path in front direction of the Unit

pchote commented:
> There is no clear barrier and meaning between Mobile and MobileInfo. There are lot of functions that Mobile could deal by himself and yet are delegated to MobileInfo...

The barrier is well defined in theory, but is perhaps not obvious and probably not implemented well in practice: `MobileInfo` (and the other trait infos) model an actor type (e.g. light tank), and `Mobile` (and other traits) model a specific actor (e.g. light tank #37, aka 'Bob').  Most of the logic in `Mobile` calls through to the Info because there isn't much actor-specific stuff in there - all light tanks behave the same.  Having it in the Info means that we can also reason about pathing without needing a specific actor to query against, which can useful in some cases.

WolfGaming commented:
The only question about this that I have for you is this, is it somehow possible to remove this move constructor public Move(IActor self, Func<List<CPos>> getPath).

Because when i was working on serialization I didnt find any way to serialize the getPath parameter.

Rydra commented:
So, the problem is the Func member itself, not the constructor, right?

Then one option could be to ignore that attribute so it's not serialized. I'm not aware if JSON can allow this kind of "ignorable" attribute, but it has to have one for sure. Look at http://stackoverflow.com/questions/1411577/ignoring-a-field-during-net-json-serialization-similar-to-xmlignore, it may help.

There is a workaround though, by using the strategy design pattern, if you still have troubles, but let's exhaust other options before that.

penev92 commented:
I'm not sure ignoring is an option, as I suspect we need **everything** to be serializeable.

Rydra commented:
@WolfGaming what is the specific error that arises when you try to serialize a class with a delegate

Rydra commented:
Dear @WolfGaming, I've implemented a special branch for you in my repo, on a branch called upstream/moverefactor-2. In this branch I get rid of the Func<List<CPos>> that was disturbing you. I invite you to try that branch and see if you can use it for your purposes.

https://github.com/Rydra/OpenRA/tree/upstream/moverefactor-2

I await any comment you may have.

Rydra commented:
After some tests, I have not altered the behaviour of the game with this refactor, so it can be reviewed and merged if necessary.

Gotta say I think this PR indicated a good way to proceed in terms of software quality and confidence, setting up tests to understand and secure the behaviour as well as to guide the architecture and detect which places can or should be refactor to ease understandibility and maintanibility.

If we could cover the code with tests a good number of functionalities, we can speed up highly the development of new features, ensure that nothing breaks and review every request waaaaaay faster.

pchote commented on OpenRA.Game/Traits/World/ActorMap.cs:
```diff
@@ -192,6 +228,9 @@ public ActorMap(World world, ActorMapInfo info)
 		}
 
+		/// <summary>
+		/// NOTE: Obsolete. Use GetActorsAt
```
`GetActorsAt` looks like an exact duplicate of this?

Rydra replied:
Exactly it is, that's why I marked the other 2 as obsolete. The problem is the return type. It's not the same to cast a IEnumerable<IActor> to IEnumerable<Actor> than IActor to Actor, I had to make changes to a whole lot of more files which would be inapproapriate for an already-big commit.

That's why I marked the other 2 as obsolete. A future PR should get rid of them.

pchote replied:
Could we implement one in terms of the other here, using `Cast<>`, for now?

Rydra replied:
Done

RoosterDragon replied:
Seeing as the interfaces are gone, we should go back to using the original methods.

Rydra replied:
Done

pchote commented on OpenRA.Game/WPos.cs:
```diff
@@ -35,6 +35,10 @@ public struct WPos : IScriptBindable, ILuaAdditionBinding, ILuaSubtractionBindin
 
+		/// <summary>
+		/// Returns the position 'a' plus a percentage (defined by mult and div) of
+		/// the distance between 'a' and 'b'
```
The key phrase here is 'linear interpolation' between points a and b.

Rydra replied:
Clarified. Done.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -133,21 +139,87 @@ static int HashList<T>(List<T> xs)
 
-		List<CPos> EvalPath()
+		// NOTE TO REMOVE: This previously used mobile.ToCell. actor.Location
+		// and mobile.ToCell are exactly the same always, and using
```
Not quite exactly the same, because `actor.Location` is a property that has additional overhead.  You could minimize that by evaluating Location once and storing it in a temporary variable.

Rydra replied:
I'll have a look at the property. In this case maybe a property inside Mobile with a better name would be a better fit. Maybe a CurrentLocation property inside mobile.

Rydra replied:
Done. Created a new property inside mobile called "CurrentLocation". It's a replica of ToCell but without that mentioned overhead. Semantics are important to me :)

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -395,6 +524,8 @@ protected override MovePart OnComplete(Actor self, Mobile mobile, Move parent)
 					parent.path.Add(nextCell.Value.First);
 				}
 
+				// Start the seconf half of cell move, that, once completed,
```
seconf

Rydra replied:
Done

pchote commented on OpenRA.Test/MoveTests.cs:
```diff
@@ -0,0 +1,376 @@
+		}
+
+		[Test]
+		public void PathAvailableAndNoCollisionsReturnFirtHalfMove()
```
Firt

Rydra replied:
Done

pchote commented on OpenRA.Test/MoveTests.cs:
```diff
@@ -0,0 +1,376 @@
+		#region Support Methods
+
+		static readonly List<CPos> DefaultPath = new List<CPos>
+													{
```
Wonky indentation

Rydra replied:
Done

pchote commented on OpenRA.Test/TestUtils.cs:
```diff
@@ -0,0 +1,108 @@
+		static WVec OffsetOfSubCell(SubCell subCell) { return subCellOffsets[(int)subCell]; }
+
+		static WVec[] subCellOffsets = new WVec[]
+									{
```
Wonky indentation.

Rydra replied:
Done

pchote commented on OpenRA.Test/MobileInfoTests.cs:
```diff
@@ -0,0 +1,95 @@
+				new CPos(1, 1), actor.Object, CellConditions.All);
+
+			Assert.IsFalse(result);
+		}
```
The interesting and subtle cases that benefit most from testing are where you have crushing, and checking against the same cell that you're already in (there known buggy behaviour in this case).

Rydra replied:
This could be a good addition for a next PR focused in testing and analyzing the behaviour of crushes. But first it needed to be refactored and understood :)

WolfGaming commented:
@Rydra Ill have a look at your branch when my summer holidays start, which would be at the latest 8th of May.

Rydra commented:
@pchote amended all your requested changes.

pchote commented:
I've put a bit of thought into this, and my earlier uneasyness over the direction we're heading with interfaces, mocks, and unit tests has solidified into a definite opinion.

Before I start my wall of text below, let me commend you on your general pathfinder/mobile reworking - that's great, and not an issue here (and I hope I don't scare you off with what follows).  It's also not about unit testing in general, but only their specific application to OpenRA.  I'm hoping that this spawns some discussion and alternative development rather than shutting down the pr!

#### 1. It isn't viable to test real-world bugs using units and mocks

The majority of our gameplay logic relies on one or both of (a) interaction between multiple traits, or (b) lambdas that run at a later time.  Both of these cases are tricky to usefully unit test.

One could argue that (a) isn't a problem because we can just mock out everything that a trait interacts with. Similarly, (b) could be tested by writing mocks that store and run these lambdas to that their results can be tested.  I think this is (at best) short sighted, as it would require a significant amount of fiddly new code which will almost certainly miss subtle details in the real traits.  It would be difficult to guarantee that the tests don't diverge further as their counterpart traits evolve.

Consider a bug like #6076 (and the related ones linked inside): this was triggered when two conflicting orders were resolved during the same tick (in this case attack and then sell).  I'm not sure where we would even begin with writing unit tests that could catch this kind of error, which is a common one and representative of the type of thing we need these tests to handle.

Another example is the commit 8676562d4712868478f95846fbdf08f8fe509647 from #3063.  This "fixed" what seemed to be an obvious bug that caused units on the (intact) sides of a destroyed bridge to explode when the bridge was repaired.  The problem was that CanEnterCell returned false for the cell the actor is already in.  The bridge code then assumes that this is a tank that has been dumped in the water (or a boat that is now beneath the newly-repaired bridge), and kills it.  If we had unit tests checking CanEnterCell then I would almost certainly have changed it, because "of course this behaviour is wrong".

Unfortunately, other parts of the code rely on that specific behaviour, and this change completely broke harvesters (#3089) and a couple of other things (which I don't recall the specifics of).  We ended up reverting that commit, and IIRC the bridge repair problem has still not been fixed.

My point there is that a unit test wouldn't have helped, but a higher level "check that harvesters still work" test would have revealed it immediately.

I'm also not saying that unit tests are completely useless for these kind of interactions, but more that the number of tests they would need and level of complexity in each to make them a useful part of our development is looking quite prohibitive (compared to the alternative).  Unit tests will also be useful at testing the core engine functionality, which use a more traditional structure.

#### 2. We have a planned alternative that should (in theory) work better

We have a plan for how to test situations like the ones above, which is to come up with a way to load a map, do some setup, run it for N ticks without a renderer, and then test the results.  The missing piece here is working out how we want to implement the initial setup - instantiating from a self-contained test file vs having a dedicated test mod with tests implemented via scripted maps (@cjshmyr wrote an example scripted test map a while ago) vs something completely different.

The tests would run specific gameplay scenarios such as "blow up a bridge with actors on it" or "operate a refinery with three harvesters for five (simulated) minutes" and check at the end that the desired actions have occurred ("actors in the middle have died, but not those at the edges" and "harvesters are still active and there are > X credits in the refinery").  This kind of test code should be easy to build and maintain once the setup mechanism has been created.

#### 3. *grumble* Industry standards, over-abstraction, buzzwords *grumble*

This one is subjective and probably controversial, but bear with me.  OpenRA has its own set of conventions and design patterns that have been influenced by a bunch of outside sources (including non-OO languages), and some of these may seem pretty weird for someone coming fresh to the code.  These have evolved over time, and work very well for OpenRA - we've received a lot of compliments about the structure of the engine from people who have gotten over the initial learning curve and started working with it.

I expect that your experience with the pathfinder and Mobile code has been rather different, as everyone agrees that corner of the code is just completely terrible.  It violates our own conventions, and most people refuse to touch it because of its opaqueness and fragility.  Given this, I wouldn't blame you if your opinion of OpenRA and its developers was "these bozos don't know anything and need to be educated about how real object oriented code is written".

From the other side of the fence, it can be tricky to diplomatically deal with pull requests of the form "rewrite system X to follow industry standard Y" because, while done with good intentions, they often implicitly claim "general-purpose rule A automatically overrides your specific use case B and experience C because that's just how these things are done".  In some cases that's true (there's certainly a decent chunk of bad code and abstractions in the codebase), but it isn't universal.

Getting back to specifics, I think that the work you've done on the pathfinder and Mobile are examples of the good and much needed kind of change, but that introducing all these new interfaces for mocking the world, map, tileset, traits, actors, etc is an example of letting the standard dogma override the project conventions to its detriment.  Given the limited utility (argued above) of unit-testing trait code, they add a lot of extra confusion, duplication, and points of failure that will hurt maintainability down the track.

To wrap up, i'm not asking for any specific changes here yet, but think this is something that needs to be discussed and decided before we proceed.  I think it would be useful to hear from @chrisforbes, @RoosterDragon, @huwpascoe who I have discussed elements of the above with in the past.

penev92 commented:
Having said all that ^ (I admit, I didn't read it, I just skimmed through it), everything movement-related in OpenRA still pretty much runs on dark magic (when it does work) and is in dire need to at least be refactored.

Rydra commented:
First, thanks @pchote for taking the time to write such a comprehensive comment about architecture. It's really fulfiling to see that something I made has created some kind of serious discussion in a community.

It's obvious that business and videogame worlds are extremely different. But you know, unit testing is not the only option that it is available for testing. In fact what you are talking about are integration tests, which tests the interaction between different "units", while Unit tests are about testing one thing in isolation. Unit tests do NOT replace the other. That's why I asked some time ago, for instance, some help regarding about how to instantiate, for instance, a Map in isolation, since the constructor is quite complex.

Having said that, unit testing is about two things: confidence (which is different than correctness) and guidance. I'm quite fond on TDD (Test-driven development) and in a slowly and steady way, unit tests show you quite well what you have to expect from the code you intend to add or change, because it testes BEHAVIOUR (Which also guides you quite well about which architecture and patterns you should use to keep stuff in isolation). If you had some tests around the code you have to change, you can know (way better than without tests) when fiddling with that piece of code whether you are breaking it or not.

They also help in UNDERSTANDING the code. Holy cow, understanding how did the pathfinder, mobile and move activity worked was hellishly challenging (yet rewarding when I made it). Without separating the code in bounded areas (through interfaces) would had made it way harder.

TL;DR#1: Unit tests help in understanding the code and test the behaviour of a piece of code in isolation. But for that to work the code needs to be TESTABLE (you'll love this: http://misko.hevery.com/attachments/Guide-Writing%20Testable%20Code.pdf). Furthermore, unit tests do not replace traditional testing, they add value by providing more confidence.

Ohhh, and it's quite common in TDD projects that there is more testing code than production code. If you see my Movetests.cs file, these tests provide a code coverage of roughly 90% of the Move class. Yet this does not mean that the code is bullet-proof.

@pchote your #2 does not enter in conflict with my PRs nor with unit tests.

Regarding your concern about me introducing more complexity with those interfaces, etc... allow me to make my points:

- Refer to the guide of writing testable code I posted above.
- As you have seen, I haven't changed the existing architecture at all. I just added a "man-in-the-middle" (the interfaces) which allowed me to test code and understand what the hell was going on. Also those interfaces allowed me to do two very important things if you remember well from the Pathfinder PR: Switch implementations easily and microbenchmarking.

I had two implementations of the pathfinder: LegacyPathSearch and PathSearch, which implemented a common interface. And the Trait only interacted with the interface, never knowing nor caring which was the implementation, and switching between legacy and current and benchmarking them was a piece-o-cake. Once I could see that the new implementation works faster and better, I could safely remove Legacy.

And you'll really be grateful of the IGraph interface once you decide to implement a hierarchical pathfinding, because the A* algorithm can be graph-agnostic as long as it fulfills the interface.

To sum up, I haven't changed the architecture, I just add a small layer between layers that decoupled the dependencies.

I hope I made my points clearer.

Mailaender commented:
Needs a rebase.

Rydra commented:
Rebased.

Rydra commented:
I would need this PR to be accepted as it would ease some unit testing I would wish to perform regarding pathfinding. Is there anything I can do to speed up the acceptance of this?

pchote commented:
I don't expect much progress here until more people weigh in on the topic.  You could try perhaps try prodding people in IRC to comment.

Like I mentioned above, I think that TDD is fine in general, and the IGraph interface was a good idea.  While there's nothing in your comment above that I specifically object to (except perhaps the idea that we would need multiple Actor implementations), I also don't think it really counters my main assertion.

My specific objection is to the idea that the specific benefits to OpenRA that come from unit testing the trait code (which I argued aren't nearly as large as one might first assume) are large enough to outweigh the increased indirection, code volume, and fragility that would come along with them.  That makes them a net-negative in the grand scheme of things.

Rydra commented:
The matter and main issue is that I'd wish to test stuff in isolation, without any noise from other classes if possible in a controlled environment. World and Player for example are environment classes and I'd wish to easily instantiate and mock them. Same goes to actor and other classes.

The problem with them is that they are hard as the hell to instantiate. I need separate complex files to do a very simple thing.

One example would be a test I have in mind regarding my other PR for improving harvester logic: search a specific square that fullfills conditions in a cellLayer. For this I have dependencies with map, world, actors... And I don't want full blown implementations of them, I just want to put my function in a test harness and see that it works.

It adds a layer (which is an interface) but the performance hit is negligible compared with the performance improvement I get because I finally could safely refactor and understand what it does.

I understand the code volume counterpart (yet I disagree with your "fragility" point).

As I said, the point of creating interfaces is to ease and make safer these refactors I'm applying and minimize regressions.

Please read this:http://www.gamasutra.com/blogs/AlistairDoulin/20111124/90671/Test_Driven_Game_Development_Experiences.php

Unit158 commented:
Honestly, the unit tests do prove useful. I am often starting a game and playing until a certain event occurs then using breakpoint to examine the execution and result of the code, which is slow as hell. The unit tests do help when I am testing an individual function and I don't know how well it works or if it works for the right reason. They aren't for everyone's problem, however.

On 2015-03-26, at 6:27 AM, David Jiménez <notifications@github.com> wrote:

> The matter and main issue is that I'd wish to test stuff in isolation, without any noise from other classes if possible in a controlled environment. World and Player for example are environment classes and I'd wish to easily instantiate and mock them. Same goes to actor and other classes.
> 
> The problem with them is that they are hard as the hell to instantiate. I need separate complex files to do a very simple thing.
> 
> One example would be a test I have in mind regarding my other PR for improving harvester logic: search a specific square that fullfills conditions in a cellLayer. For this I have dependencies with map, world, actors... And I don't want full blown implementations of them, I just want to put my function in a test harness and see that it works.
> 
> It adds a layer (which is an interface) but the performance hit is negligible compared with the performance improvement I get because I finally could safely refactor and understand what it does.
> 
> I understand the code volume counterpart (yet I disagree with your "fragility" point).
> 
> As I said, the point of creating interfaces is to ease and make safer these refactors I'm applying and minimize regressions.
> 
> Please read this:http://www.gamasutra.com/blogs/AlistairDoulin/20111124/90671/Test_Driven_Game_Development_Experiences.php
> 
> —
> Reply to this email directly or view it on GitHub.
>

Rydra commented:
I've added a new commit that merges with #7658. In the PathfinderTests.cs file I added a new Test for the pathfinding Search algorithm to prove you, @pchote, how easy it's been for me to unit test it. This test will also tell me visually if the call is too complex (which indeed it is) and give me a clearer idea how to tackle it.

Not only this, but now given that there are no heavy dependencies, I can easily benchmark with Stopwatches precisely the inner algorithm of Search and see how changes to this algorithm (because the current implementation is not the most optimal, that's to say) affects execution time.

RoosterDragon commented:
I'm going to be that annoying guy and say I generally agree with both of you.

I agree with @pchote in as far as throwing interfaces and abstractions at most of the traits won't likely gain us much as the bugs tend to occur at the integration level and not the unit level.

That said, I think the pathfinding and movement is something that would benefit significantly from being detangled from the rest of the code. It's so directly interdependent on so many things that nobody has been really willing to do anything substantial with it until now.

Anyway, to get to the point about getting the commit reviewed. Try breaking it down into a bunch of smaller, easily consumable, commits or even PRs. ~2500 changes is simply too much for most of us to work up the motivation to go through in one go, I expect.

huwpascoe commented:
I found it impossibly hard to write tests for the project when installing NUnit, because everything is so tightly connected. You can’t even create an actor without a world. “yeah yeah, actors don’t exist outside of the world bro.” But that’s the problem. What purpose other than mistrusting imaginary programmers does that strange entangled structure exist? There’s no reason for it.

So here’s someone who’s experience and clearly knows what he’s doing, at least more than most of us, who’s put so much work into really fixing this long term issue. After taking him through the journey of coding style and rejected and rebased PRs, you’re finally at the point where it’s time to take the leap of faith and you’re too scared to jump.

What are you waiting for? The next release deadline is never; the master branch is still very much a mess. Merge it, then fix things when they break. No commit is noise! Bugs happen! What do you want?! : P

Rydra committed:
Extended interfaces Actor, World and Map to support future changes on
movement and pathfinding and ease Testability

Rydra committed:
Refactored MobileInfo, Mobile and Move to make them testable in isolation, easier to understand what they do by commenting and restructuring the code making sure to not cause undesired side-effects.

The fact of having to break dependencies on a highly coupled system forced me to modify and create several interfaces, that at the same time forced me to change several classes directly and indirectly related to Move, Mobile and MobileInfo.

Rydra committed:
Added Unit Tests to cover several functionalities related to movement,
graph processing and pathfinding.

More tests could be added in the future, in addition to integration tests,
increased coverage and reorganization. At least foundations have been
stablished

Rydra commented:
As suggested by @RoosterDragon, I've reorganized the PR into 3 discrete, separate commits. The first one prepares the infrastructure that allows me to refactor and reorganize Mobile and Move, the second one does the refactor, and the third one adds unit tests to cover them.

As discussed, unit tests DO NOT cover all possibilities, yet they help a lot in reducing the chance of bugs and creating code which is easier to understand.

This PR also merges #7658.

Mailaender commented:
@chrisforbes What is your opinion on the matter?

pchote commented:
@huwpascoe: I believe I have covered your main points in my comment above: specifically my concerns about return on (code) investment, and dogma overriding specific project requirements.

huwpascoe commented:
I finished with 'what do you want?!' (in my usual hyperbolic style) because I really mean that. How do you see the project? What is the goal of the project?

Are your decisions in favor of the goal?

Mailaender commented:
Can we please return to an objective debate?

huwpascoe commented:
@Mailaender

Well if you insist. Here's the debate in a nutshell:

For:
This code improves test-ability and reduces spaghetti code.

Against:
I don't see any obvious improvement, ergo, there's no in changing what ain't a problem.*

(read: it's not a bug it's a feature)

huwpascoe commented:
Come to think of it, if you didn't even want unit tests, why'd you bother agreeing to have NUnit installed 
in the first place?

obrakmann commented:
We do want unit tests. But there are things where unit tests are suitable, and others where they are not. A recent example: [These changes](https://github.com/OpenRA/OpenRA/pull/7748/files#diff-ad4437e58b6bba75e4be119fc69224c3R715) to `Map::CellContaining()` and `Map::CenterOfCell()` would be easily testable using unit tests.

The Move activity, however, is not such a case.  Its behaviour depends on far too many factors: timing, inter-dependence and interaction with other traits, activities and their respective states, the order of orders a unit receives, the state of the world and so on.  This makes it untestable to me. You can probably test some things, but these will be just the easiest cases that a proper review will catch just as well. If you want to create a really comprehensive test suite for the Move activity, the effort required would be huge and will probably never be complete, and I'd say it's just not worth it.

To summarize: Refactoring the Move activity: yes, please! But unit tests are not worth it and would just become a huge maintenance burden.

Rydra commented:
@obrakmann Move activity can and should be unit testable, even if it is for the sake of detecting poor design smells. As I was moving code around, here and there, I could see at first sight one problem: dependency injection and demeter laws violation.

Not only this, but just to name another example of dependency problem, actor has a property, "World", which tells in which world the actor lies. If World is a singleton and already contains structures of its actors, what's the point of coupling even further? I'm not going further into this example because it's silly, but I hope I made my point.

Another clear example, Mobile trait, which is a dependency used across other classes, has some semantics issues. Pathfinder (pathsearch to be more specific) requires a MobileInfo instance to work even for a harvester who wants to look for a harvestable cell. What if I want a turret to look for an enemy target using such algorithms (I wanted to test this but you know... turrets do not have MobileInfo trait...).

And more etcetcetc...

Move activity has around 5 to 6 dependencies with some cleverness, could be niftly reduced. And unit tests help me detect these smells and flaws because they make me ask the following question: If I were new and I'd wish to develop this functionality,  As a client/user of that class/function, HOW WOULD I WISH TO CALL THIS FUNCTIONS(Parameters, name, etc...)/CREATE THESE CLASSES (Note I'm implicitly declaring behaviour and interfaces when questioning this). If the answer to this question is not in sync with the current implementation... then we've got a problem probably. This is the case I found with, for instance, MobileInfo and Pathfinding.

Hey, that comes to my mind... Do I really need an actor to perform a pathfinding? Maybe an "IMobile/IMovable" (an object that moves) could be enough... (just a simple example about how unit tests may help an enhance thinking process).

And they do not suppose a maintainability burden, because you can always delete them. I don't think it's a good idea of thinking about tests as a burden instead of a safe-guard.

huwpascoe commented:
@obrakmann So tests are okay, but only for features that won't break. Okay then...

Unit158 commented:
:-1: for how the tests are designed in this case. IActor and IWorld (etc, I haven't read the entire thing)  should not be needed. There is no value outside of tests that benefit from the extra abstraction. 

Perhaps functions should be extracted from these methods to make them testable, but adding unnecessary abstraction is a giant code smell for video game design and any other high speed or low resource environment.

On 2015-03-29, at 5:23 PM, yesfish <notifications@github.com> wrote:

> @obrakmann So tests are okay, but only for features that won't break. Okay then...
> 
> \
> Reply to this email directly or view it on GitHub.
>

huwpascoe commented:
"but adding unnecessary abstraction is a giant code smell for video game design and any other high speed or low resource environment."

Hmm... you have a point there. The extra MOV instructions required to offset the IVTable are orders of magnitude more CPU intensive than the massive tangled web of system.reflection calls, object factories and linq extensions that make up the core of the game, which is totally [not still broken](https://github.com/OpenRA/OpenRA/issues/6818) anyway.

Rydra commented:
"There is no value outside of tests that benefit from the extra abstraction."

That's some hefty value...

Unit158 commented on OpenRA.Test/PathfinderTests.cs:
```diff
@@ -63,84 +83,91 @@ static bool IsValidPos(CPos pos, int mapWidth, int mapHeight)
-				path8 = pathfinder.FindBidiPath(search, search2);
-			}
+		[Test]
+		[Ignore("Not implemented yet")]
```
Unfinished code should not be submitted.

Unit158 commented on OpenRA.Test/MobileInfoTests.cs:
```diff
@@ -0,0 +1,95 @@
+		}
+
+		/// <summary>
+		/// If actor doesn't share cell and there's not a free subcell, KO
```
KO?

Unit158 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -31,44 +32,173 @@ public enum CellConditions
+		/// <summary>
+		/// Determines whether the actor is blocked by other Actors
+		/// </summary>
+		/// <param name="world">The world where the actor moves</param>
```
These shouldn't need description.

Unit158 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -31,44 +32,173 @@ public enum CellConditions
+		/// <summary>
+		/// Determines whether the actor is blocked by other Actors
+		/// </summary>
+		/// <param name="world">The world where the actor moves</param>
```
These shouldn't need description.

Unit158 commented:
@huwpascoe Tu quoque? This doesn't fix those problems, so there is no point in addressing them here.

@Rydra I like the work you've done, but if the constructors are too complex to do whatever tests you need to, that is a problem that should be fixed first. For me, adding this layer of abstraction feels like adding a giant band-aid over something else that should be fixed. I may not understand the problem fully, but it makes no sense in my mind to change such a large portion of the code to use interfaces to accommodate one side of the story, which isn't the side of the story that the users care about.

If you can elaborate on why you need these interfaces and why it isn't possible to fix the base classes, I'll be convinced.

huwpascoe commented:
@Unit158 Yes! (I'm glad one person's listening at least) Any step is improvement over the status quo at this point, because once the path-finding is properly implemented, then you have the grounds to decide whether or not interfaces belong and remove them in a future PR. But to stifle progress now in the name of...um...temporary code purity? is a dumb thing.

Unit158 commented:
In any case, this should not be merged until the next release is finished. We have a while to decide yet.

Rydra commented:
@Unit158 The overall necessity (to me at least) for requiring those interfaces can be reduced easily to code analysis and understanding: 

- Unit testing marks explicitly which is the current behaviour of the code and how it is used by other parts of the code, and gives me hints about the function signatures, constructors and overall architecture flaws. Then these unit test help me in that, whenever I move code around to refactor and make it more understandable, I don't break many things.

Let's look at the most recent case, the Pathfinder. Take a look at the Pathfinder trait in master and compare it to the current bleed. Quite different, right? I refactored all the tangled code into smaller, separate and understandable units or concepts that can be analyzed separately and reused if necessary (the Graph abstraction comes to mind). How do you think I came up with all those classes out of an initially single class called "PathSearch"? It took more than a month to figure things out, and unit tests made this easier. While unit testing and refactoring the mind goes into a "creative&sensing state" (or at least it is how I feel it) to detect smells.

IActor, IWorld, IMap, etc... are necessary because the Pathfinder depended on them but hell, I'm not interested in refactoring nor testing the behaviour of these classes, I'm exclusively interested in testing and refactoring the pathfinder without noise from other classes. Thanks to this, I can really focus on what I'm interested. Actor and World are complex objects and I don't really wanna go through the fud of having to create a MiniYaml because I just want to test another class.

There's a saying: "Early optimization is the root of all evil". And it's quite true. @pchote has claimed that these interfaces introduce indirection, etcetcetc... and while it's true that calling a class through an interface takes ~5 ns more (ref: https://msdn.microsoft.com/en-us/library/ms973852.aspx) the code first needs to be "in shape" before these kind of optimizations.

Besides, traits and all kind of stuff around the code uses things like IMove, IDumb, IWhatever, etc... can't really understand the alarmism for a mere interface i introduced. In fact with some cleverness I could minimize things like "IMobile", etc by making a composition of those interfaces (a thing I'm already partially doing). That's called "Interface segregation" and it's one of the SOLID Principles (The "I" in SOLID).

penev92 commented:
I'm all for "Off-Topic" btw.

pchote commented:
@Rydra: In my comments above I was referring primarily to code development rather than CPU pointer chasing.  To simple examples: hiding all of the core gameplay classes behind interfaces means that you have to modify two places instead of one when adjusting their interface, and impacts the usefulness of the "go to definition" IDE feature.

You mentioned in IRC that resharper makes it trivially easy to add and remove these kind of interfaces.  Have you considered keeping them on your development branches so that you can use your favourite development strategies, and then removing them from your final PRs?  That way we can avoid these arguments and focus on development instead.

Rydra commented:
@pchote I could do what you suggest at the expense of removing all the unit test which would make me feel sad and people would be unable to see how the code I implemented works... I'll consider it, but let's keep going this discussion, IMHO it's interesting. Actually the process of deleting an interface just like IActor consists in removing the interface and replace the text "IActor" to "Actor", "IWorld" to "World", etcetcetc... No need really for resharper.

Btw regarding the "go to definition" (shortcut F12 IIRC), if you use Ctrl + F12 it brings you to the implementation instead of the interface. And having to modify 2 places instead of one you can right click over the signature you want to change -> Refactor -> Change signature/Rename/whatever.

chrisforbes commented:
My position here is that forcing a seam with interfaces *just* so that you can mock them is pretty much the worst way to untangle code. Sometimes it's the only thing you can reasonably do, but we should always be aware that mocks are a crutch for crap OO code that wanders all over the object graph, not a good design technique.

The better abstraction here would be to isolate the *data* which is required for the pathfinding to operate. Extract that, eagerly (or lazily if you absolutely must) before starting to pathfind, and adjust the pathing code to refer only to that blob of data. Have it produce a similar blob of results.

This produces a better seam:
- It's very clear what the pathing depends on, and what it doesn't
- You can test it trivially without needing to care about the game's object graph madness at all.
- If you've done it sensibly, the pathing itself will exhibit much better cache behavior, which is the "whole game" in performance, not this nonsense about timings of virtual calls.
- You get halfway to being able to schedule the actual pathing effort on a worker

Rydra commented:
@chrisforbes when you say "isolate the data which is required..." do you mean creating some kind of DTO passed to the pathfinder that contains simple data like integer, strings whatever instead of the proper actor, world, mobile etc... Classes?

I don't completely agree with you saying that it is the worst way to untangle code, lots of modules will inevitable have always dependencies and work together to achieve a goal, that's a fact that cannot be denied.

Rydra commented:
Althought gotta say I wouldn't completely dislike the idea of some kind of "request object".

Rydra committed:
Interfaces removed

Rydra commented:
Well, I finally give up. I've removed all the interfaces (and therefore all the unit tests) and left everything with the changes I made, but as if the interfaces never existed.

I hope you like it now more...

RoosterDragon commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -34,103 +45,118 @@ public FindResources(CPos avoidCell)
+				{
+					if ((avoidCell.HasValue && loc == avoidCell.Value) ||
+						(loc - self.Location).LengthSquared > searchRadiusSquared)
+						return int.MaxValue;
```
Is this supposed to be ```InvalidCell``` now?

Rydra replied:
True. Done

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -59,70 +60,42 @@ public Move(Actor self, CPos destination, int nearEnough)
 		}
 
-		public Move(Actor self, Func<List<CPos>> getPath)
+		private void Init(Actor self, CPos? destination, WRange nearEnough)
```
Looks like you could make this a private ctor and chain calls to it, which would allow you to keep the readonly modifiers you had to drop.

Rydra replied:
The problem is that there already exists a constructor with the same signature: Move(Actor self, CPos destination, WRange nearEnough). I making that init a private ctor and they collide.

RoosterDragon replied:
Ah good point - in that case just drop the ```private``` keyword.

Rydra replied:
Done

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -168,37 +145,38 @@ public override Activity Tick(Actor self)
 				return this;
 
-			var firstFacing = self.World.Map.FacingBetween(mobile.FromCell, nextCell.Value.First, mobile.Facing);
+			// Check if a turn is required. If it does,
```
Maybe ```If it does``` -> ```If so```.

Rydra replied:
done

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -233,39 +234,40 @@ void SanityCheckPath(Mobile mobile)
-					return null;
+				{
+					return false;
+				}
```
No need to introduce braces.

Rydra replied:
done

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -283,21 +285,23 @@ public override IEnumerable<Target> GetTargets(Actor self)
-		abstract class MovePart : Activity
+		// NOTE: This class is uncancellable until it completes its intended move...
+		// If we could solve this, we could enable path smoothing!
+		public abstract class MovePart : Activity
```
Doesn't need public any more, as far as I can tell.

Rydra replied:
Never been a fan of internal private classes. That makes them just harder to understand from the outside. Besides, I have to say I did that because I planned to get rid of them sooner or later once I could separate responsibilities with tests. I would not make them private again, but if they hurt your eyes then I'll make them back to private.

RoosterDragon replied:
These just feel like implementation details that shouldn't be publicly exposed to me. I'll let somebody else weight in.

pchote replied:
I agree with @RoosterDragon.

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs:
```diff
@@ -98,7 +98,7 @@ public override Activity Tick(Actor self)
 			{
 				// Target became invalid. Move to its last known position.
-				Target = Target.FromCell(self.World, targetPosition);
+				Target = OpenRA.Traits.Target.FromCell(self.World, targetPosition);
```
This qualification seems uneeded?

Rydra replied:
done

RoosterDragon commented on OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs:
```diff
@@ -131,7 +131,7 @@ public override IEnumerable<Target> GetTargets(Actor self)
 				return inner.GetTargets(self);
 
-			return Target.None;
+			return OpenRA.Traits.Target.None;
```
Also here.

Rydra replied:
Done

RoosterDragon commented on OpenRA.Mods.Common/Pathfinder/PathSearch.cs:
```diff
@@ -46,13 +50,19 @@ public static IPathSearch FromPoint(IWorld world, IMobileInfo mi, IActor self, C
+			{
+				var locInfo = search.Graph[loc];
+				return locInfo.EstimatedTotal - locInfo.CostSoFar == 0;
+			};
```
Maybe have ```PathSearch``` to pass ```this``` as a parameter to the ```isGoal``` delegate? Seems odd to need to capture the search object so you can query the graph but maybe I'm being overly specific to this use-case.

Rydra replied:
It's just a factory constructor so it's not a big deal. Besides, from a behavioural and reusability aspect, the isGoal delegate should only know about location as a parameter (for searches and stuff like that). That delegate could be perfectly placed as a private method as it was before, but I consider it pointless).

RoosterDragon replied:
That's fair.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -110,7 +110,7 @@ bool IsAcceptableProcType(Actor proc)
 		}
 
-		Actor ClosestProc(Actor self, Actor ignore)
+		public Actor ClosestProc(Actor self, Actor ignore)
```
Doesn't need public any more.

Rydra replied:
That's something I was considering, yet given the weight that this function has and the usefulness that it may have, and how specific its specification is (get the closest Proc for a given actor) I considered it would be useful from a testing standpoint to keep it public and available to other areas in case of necessity.

RoosterDragon replied:
OK.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -103,50 +125,34 @@ TerrainInfo[] LoadTilesetSpeeds(TileSet tileSet)
+		/// This constructor is merely used for testing since I can't
+		/// assign nor mock readonly public fields.
+		/// </summary>
+		public MobileInfo(bool sharesCell, string[] crushes)
```
I guess we don't need this now?

Rydra replied:
Unfortunately, true... Removed.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -303,23 +275,59 @@ public bool CanEnterCell(World world, Actor self, CPos cell, out int movementCos
 
-		[Sync] public int Facing
+		[Sync]
+		public int Facing
```
Don't commit this formatting change (it'll be inconsistent with the rest of the codebase).

Rydra replied:
Done

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -331,6 +339,9 @@ public class Mobile : IIssueOrder, IResolveOrder, IOrderVoice, IPositionable, IM
 		[Sync] public CPos ToCell { get { return toCell; } }
 
+		// A property created for the sake of semantics and understanding
+		[Sync] public CPos CurrentLocation { get { return toCell; } }
```
This won't need to be marked as [Sync].

Rydra replied:
Done

pchote replied:
This is a duplicate of `Actor.Location`.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -419,9 +430,10 @@ public void SetVisualPosition(Actor self, WPos pos)
-			self.World.ActorMap.AddInfluence(self, this);
-			self.World.ActorMap.AddPosition(self, this);
-			self.World.ScreenMap.Add(self);
+			var actor = self;
```
Is there a reason for this local variable?

Rydra replied:
Quite strange I did that, I wonder what I was thinking on.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -777,7 +795,8 @@ public Activity VisualMove(Actor self, WPos fromPos, WPos toPos, CPos cell)
 
 			var facing = Util.GetFacing(toPos - fromPos, Facing);
-			return Util.SequenceActivities(new Turn(self, facing), new Drag(self, fromPos, toPos, length));
+			var actor = self;
```
Also this local?

Rydra replied:
Done

RoosterDragon commented on OpenRA.Game/GameRules/ActorInfo.cs:
```diff
@@ -16,10 +16,15 @@
 
 namespace OpenRA
 {
+	public interface IActorInfo
```
I guess this interface can also go now.

RoosterDragon commented on OpenRA.Game/Map/ActorInitializer.cs:
```diff
@@ -14,10 +14,26 @@
 namespace OpenRA
 {
-	public class ActorInitializer
+	public interface IActorInitializer
```
This one too, unless you did have a plan for these ones that you've kept around.

Rydra replied:
The two interfaces you pointed out skipped my hammer :). Removed!

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -183,78 +179,60 @@ public int TileSetMovementHash(TileSet tileSet)
+			return false;
+		}
+
+		public bool CanEnterCell(World world, Actor self, CPos cell, out int movementCost, Actor ignoreActor = null, CellConditions check = CellConditions.All)
```
Could we move this next to its cousin that lack the out parameter, and reimplement the cousin to call this and just drop the out parameter on the floor to avoid duplication.

Rydra replied:
Done

RoosterDragon commented on OpenRA.Game/Traits/World/ActorMap.cs:
```diff
@@ -25,7 +25,43 @@ public class ActorMapInfo : ITraitInfo
 	}
 
-	public class ActorMap : ITick
+	public interface IActorMap
```
Also this interface :)

Rydra replied:
Done

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -13,13 +13,14 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
+	#region Interfaces and Enums
```
No need for this region now.

Rydra replied:
Done

obrakmann commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -34,103 +45,118 @@ public FindResources(CPos avoidCell)
+		/// Finds the closest harvestable pos between the current position of the harvester
+		/// and the last order location
+		/// </summary>
+		CPos? ClosestHarvestablePos(Actor self)
```
Something in here regressed, so that (some) of the harvesters in the intervention mission stop working after dropping off their first load..

Rydra replied:
I could hunt and solve the issue. Thanks!

reaperrr commented:
The shellmaps work fine, but I get a crash when trying to start a skirmish game (tested RA and TD mods). Seems to be related to crushing.
exception.log: http://pastebin.com/stYiKjnj

reaperrr commented:
@Rydra: Now that the release is out and people have more time for reviewing/testing again, any chance you might resume working on this? It would be great if this could be finished successfully.

At the very least, this would need a fix for the issue @obrakmann pointed out and of course the crash I encountered, and a rebase.

Rydra commented:
Right now I don't have much time to spare, yet I'll try to keep an eye on the couple of issues that have arised, solve them and do as @reaperrr suggests.

pchote commented:
The mix of additions, refactorings, and removals make this very difficult to review in its current state.  I've split out #8071 to hopefully start some momentum towards fixing this.

Mailaender commented:
Needs a rebase now.

penev92 commented:
Please give us a sign that you are still around, @Rydra :)

Rydra commented:
@penev92 I'm alive. A bit busy, but alive. I'll rebase and review the crush issue ASAP.

Rydra commented:
Rebase done.

@reaperrr I've been unable to reproduce your error in RA mod, in a 10 bot map for 15 min game. Could you provide me with some steps to reproduce your error?

reaperrr commented:
It's possible that it happened in a different mod. I don't remember where exactly I got that error, so I'll give it another test run to see if I can reproduce it again.

reaperrr commented:
@Rydra: Start a skirmish game with light or heavy start unit support instead of MCV only. The crash happens always if you do that, on all mods.

My suspicion, since it seems to be related to spawning start units: with your changes, some actors surrounding the MCV are spawning on top of each other.

Exception of type `System.NullReferenceException`: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei OpenRA.Mods.Common.Traits.MobileInfo.<>c__DisplayClassa.<CanBeCrushedBy>b__9(ICrushable crushable)
   bei System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   bei OpenRA.Mods.Common.Traits.MobileInfo.CanBeCrushedBy(Actor crusher, Actor actor)
   bei OpenRA.Mods.Common.Traits.MobileInfo.Collides(Actor self, Actor actor, Actor ignoreActor, Boolean canIgnoreMovingAllies)
   bei OpenRA.Mods.Common.Traits.MobileInfo.CanMoveFreelyInto(World world, Actor self, CPos cell, Actor ignoreActor, CellConditions check)
   bei OpenRA.Mods.Common.Traits.MobileInfo.CanEnterCell(World world, Actor self, CPos cell, Int32& movementCost, Actor ignoreActor, CellConditions check)
   bei OpenRA.Mods.Common.Traits.MobileInfo.CanEnterCell(World world, Actor self, CPos cell, Actor ignoreActor, CellConditions check)
   bei OpenRA.Mods.Common.Traits.SpawnMPUnits.<>c__DisplayClass6.<SpawnUnitsForPlayer>b__3(CPos c)
   bei System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   bei System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   bei System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   bei OpenRA.Exts.Random[T](IEnumerable`1 ts, MersenneTwister r, Boolean throws) in e:\OpenRA_work_clean_20150501\OpenRA.Game\Exts.cs:Zeile 151.
   bei OpenRA.Exts.Random[T](IEnumerable`1 ts, MersenneTwister r) in e:\OpenRA_work_clean_20150501\OpenRA.Game\Exts.cs:Zeile 140.
   bei OpenRA.Mods.Common.Traits.SpawnMPUnits.SpawnUnitsForPlayer(World w, Player p, CPos sp)
   bei OpenRA.Mods.Common.Traits.SpawnMPUnits.WorldLoaded(World world, WorldRenderer wr)

Rydra commented:
@reaperr cool, finally I could see the error. Let's see how I deal with it.

Rydra commented:
The error is found in the class "SpawnMPUnits" (File SpawnMPUnits.cs, line 59), It is sending a null value for the actor. This class was outside of my domain, but let's see how I can solve the issue.

Rydra commented:
@reaperrr I've managed to solve the issue. Please reset and try again. I think I've covered all the issues described in this thread.

Thanks for the feedback!

reaperrr commented:
I can confirm the crash is gone.

What I noticed is that actors which share cells (infantry) no longer start on random subcells now (unlike on bleed).

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -207,25 +185,48 @@ void SanityCheckPath(Mobile mobile)
+		bool IsInRange()
+		{
+			var cellRange = nearEnough.Range / 1024;
+			return (mobile.CurrentLocation - destination.Value).LengthSquared <= cellRange * cellRange;
```
Calculating range checks in cell coordinates is a bad idea.  I suggest keeping this in WPos / WRange, which has the numerical precision to behave properly for arbitrary diagonals.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -360,7 +363,10 @@ public override IEnumerable<Target> GetTargets(Actor self)
 
-		class MoveFirstHalf : MovePart
+		/// <summary>
+		/// Performs the visual movement from a world position of a cell to another
```
Move*Half updates the actor CenterPosition, which is the primary actor position in the engine.  That is far more than just visual movement.

reaperrr commented:
> What I noticed is that actors which share cells (infantry) no longer start on random subcells now (unlike on bleed).

Ignore that, I misremembered. There's no difference to bleed.

obrakmann commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -315,12 +317,13 @@ public override void Queue(Activity activity)
 
 			public override Activity Tick(Actor self)
 			{
+				var mobile = self.Trait<Mobile>();
```
You should probably cache this as well in the constructor. All the callers are in Move.cs anyway, and `Move` has it cached already has well, so it can just pass it in.

obrakmann commented:
In-game the changes are mostly fine. In d2k I noticed that the harvesters would often stop harvesting at some point and just sit in the spice field. I didn't see that in bleed.

The code changes look fine besides the mentioned points.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -390,6 +400,8 @@ protected override MovePart OnComplete(Actor self, Mobile mobile, Move parent)
 							Util.GetNearestFacing(mobile.Facing, self.World.Map.FacingBetween(mobile.ToCell, nextCell.Value.First, mobile.Facing)),
 							moveFraction - MoveFractionTotal);
 
+						// These two lines are doing exactly the same as mobile.SetPosition, except for setting the Visual Position.
```
What is the problem here?  The position and cell location are conceptually different things, and get updated at different times.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -406,20 +420,28 @@ protected override MovePart OnComplete(Actor self, Mobile mobile, Move parent)
-		class MoveSecondHalf : MovePart
+		/// <summary>
+		/// This class does the second part of a movement. Basically, it notifies
+		/// any possible crush in the cell once movement is terminated.
```
I suggest rewording this a bit.  The current wording seems to imply that this *only* handles that notification, which isn't true.

pchote commented:
This is a big chunk of changes to review in one go, so I would prefer it if we could first start with the smaller subset of code in #7658 (which appears to make up a chunk of the changes in this pr).

penev92 commented:
I am closing this in an attempt to clean up the PR queue a bit, until #7658 is dealt with.
Please reopen this once that is merged. :)
